runtime/src/accounts_background_service.rs

```diff
@@ -571,131 +572,141 @@ impl AccountsBackgroundService {
-                    // slot `M` satisfies `M < N`
-                    //
-                    // Proof: Assume for contradiction that we find a snapshot request for slot `N` here,
-                    // but cleanup has already happened on some slot `M >= N`. Because the call to
-                    // `bank.clean_accounts(true)` (in the code below) implies we only clean slots `<= bank - 1`,
-                    // then that means in some *previous* iteration of this loop, we must have gotten a root
-                    // bank for slot some slot `R` where `R > N`, but did not see the snapshot for `N` in the
-                    // snapshot request channel.
-                    //
-                    // However, this is impossible because BankForks.set_root() will always flush the snapshot
-                    // request for `N` to the snapshot request channel before setting a root `R > N`, and
-                    // snapshot_request_handler.handle_requests() will always look for the latest
-                    // available snapshot in the channel.
-                    //
-                    // NOTE: We must wait for startup verification to complete before handling
-                    // snapshot requests.  This is because startup verification and snapshot
-                    // request handling can both kick off accounts hash calculations in background
-                    // threads, and these must not happen concurrently.
-                    let snapshot_handle_result = bank
-                        .is_startup_verification_complete()
-                        .then(|| {
-                            request_handlers.handle_snapshot_requests(
-                                test_hash_calculation,
-                                non_snapshot_time,
-                                &exit,
-                            )
-                        })
-                        .flatten();
-                    if snapshot_handle_result.is_some() {
-                        last_snapshot_end_time = Some(Instant::now());
-                    }
-
-                    // Note that the flush will do an internal clean of the
-                    // cache up to bank.slot(), so should be safe as long
-                    // as any later snapshots that are taken are of
-                    // slots >= bank.slot()
-                    bank.flush_accounts_cache_if_needed();
-
-                    if let Some(snapshot_handle_result) = snapshot_handle_result {
-                        // Safe, see proof above
+            .spawn({
+                let is_running = is_running.clone();
+                let stop = stop.clone();
+
+                move || {
+                    info!("AccountsBackgroundService has started");
+                    let mut stats = StatsManager::new();
+                    let mut last_snapshot_end_time = None;
+
+                    loop {
+                        if exit.load(Ordering::Relaxed) || stop.load(Ordering::Relaxed) {
```

Comment:
This looks pretty good to me. My only question is: if we stop in the middle, would this leave anything broken on disk so that we can't really generate snapshot later?